File: src/Data/Long/FFI.js
var L = require("long");

// Constants
exports.zero     = L.ZERO;
exports.one      = L.ONE;
exports.negOne   = L.NEG_ONE;
exports.uzero    = L.UZERO;
exports.uone     = L.UONE;
exports.maxValue = L.MAX_VALUE;
exports.maxUnsignedValue = L.MAX_UNSIGNED_VALUE;
exports.minValue = L.MIN_VALUE;

// Utilities
exports.isLong = L.isLong;
exports.fromBits = L.fromBits;
exports.fromBytes = L.fromBytes;
exports.fromBytesLE = L.fromBytesLE;
exports.fromBytesBE = L.fromBytesBE;
exports.fromInt = L.fromInt;
exports.fromNumber = L.fromNumber;
exports.fromString = L.fromString;
exports.fromValue = L.fromValue;

// Methods
exports.add = function (l) { return l.add; };
exports.and = function (l) { return l.and; };
exports.compare = function (l) { return l.compare; };
exports.divide = function (l) { return l.divide; };
exports.equals = function (l) { return l.equals; };
exports.getHighBits = function (l) { return l.getHighBits(); };
exports.getHighBitsUnsigned = function (l) { return l.getHighBitsUnsigned(); };
exports.getLowBits = function (l) { return l.getLowBits(); };
exports.getLowBitsUnsigned = function (l) { return l.getLowBitsUnsigned(); };
exports.getNumBitsAbs = function (l) { return l.getNumBitsAbs(); };
exports.greaterThan = function (l) { return l.greaterThan; };
exports.greaterThanOrEqual = function (l) { return l.greaterThanOrEqual; };
exports.isEven = function (l) { return l.isEven(); };
exports.isNegative = function (l) { return l.isNegative(); };
exports.isOdd = function (l) { return l.isOdd(); };
exports.isPositive = function (l) { return l.isPositive(); };
exports.isZero = function (l) { return l.isZero(); };
exports.lessThan = function (l) { return l.lessThan; };
exports.lessThanOrEqualTo = function (l) { return l.lessThanOrEqualTo; };
exports.modulo = function (l) { return l.modulo; };
exports.multiply = function (l) { return l.multiply; };
exports.negate = function (l) { return l.negate(); };
exports.not = function (l) { return l.not(); };
exports.notEquals = function (l) { return l.notEquals; };
exports.or = function (l) { return l.or; };
exports.shiftLeft = function (l) { return l.shiftLeft; };
exports.shiftRight = function (l) { return l.shiftRight; };
exports.shiftRightUnsigned = function (l) { return l.shiftRightUnsigned; };
exports.rotateLeft = function (l) { return l.rotateLeft; };
exports.rotateRight = function (l) { return l.rotateRight; };
exports.subtract = function (l) { return l.subtract; };
exports.toBytes = function (l) { return l.toBytes(); };
exports.toInt = function (l) { return l.toInt(); };
exports.toNumber = function (l) { return l.toNumber(); };
exports.toString = function (l) { return l.toString; };
exports.toUnsigned = function (l) { return l.toUnsigned(); };
exports.xor = function (l) { return l.xor; };

File: src/Data/Long/FFI.erl
-module(data_long_FFI@foreign).

-export([ add/1
        , 'and'/1
        , compare
        , divide/1
        , equals/1
        , divide/1
        , equals/1
        , getHighBits/1
        , getHighBitsUnsigned/1
        , getLowBits/1
        , getLowBitsUnsigned/1
        , getNumBitsAbs/1
        , greaterThan/1
        , greaterThanOrEqual/1
        , isEven/1
        , idOdd/1
        , isPositive/1
        , isZero/1
        , lessThan/1
        , lessThanOrEqual/1
        , modulo/1
        , multiply/1
        , negate/1
        , 'not'/1
        , notEquals/1
        , 'or'/1
        , shiftLeft/1
        , shiftRight/1
        , shiftRightUnsigned/1
        , rotateLeft/1
        , rotateRight/1
        , subtract /1
        , toBytes/1
        , toInt/1
        , toNumber/1
        , toSigned/1
        , toString/1
        , toUnsigned/1
        , 'xor'/1
        ]
  )

add(A) -> fun(B) -> A + B end.
and(A) -> fun(B) -> A band B end.
compare(A) -> fun(B) when A > B -> 1;
                 (B) when A == B -> 0;
                 (_) -> -1
              end.

divide(A) -> fun(B) -> A div B end.
equals(A) -> fun(B) -> A == B end.

getHighBits(A) -> A bsr 32.

getHighBitsUnsigned(A) -> A bsr 32.

getLowBits(A) -> A band 16#ffffffff.

getLowBitsUnsigned(A) -> A band 16#ffffffff.


getNumBitsAbs(A) -> getNumBitsAbs(A, 0).

getNumBitsAbs(A, N) when A == 0 -> N;
getNumBitsAbs(A, N) -> getNumBitsAbs(A bsr 1, N + 1).

greaterThan(A) -> fun(B) -> A > B end.
greaterThanOrEqual(A) -> fun(B) -> A >= B end.
isEven(A) when A >= 0 -> (A band 1) == 0.
isOdd(A) when A > 0 -> not isEven(A).
isPositive(A) -> A > 0.
isZero(A) -> A == 0.
lessThan(A) -> fun(B) -> A < B end.
lessThanOrEqual(A) -> fun(B) -> A =< B end.
modulo(A) -> fun(B) -> A rem B end.
multiply(A) -> fun(B) -> A * B end.
negate(A) -> -A.
not(A) -> bnot A.
notEquals(A) -> fun(B) -> A /= B end.
or(A) -> fun(B) -> A bor B end.
shiftLeft(A) -> fun(B) -> A bsl B end.
shiftRight(A) -> fun(B) -> A bsr B end.

shiftRightUnsigned(A) -> fun(B) -> A bsr B end.

rotateLeft(A) -> fun(Num) ->
  ((A bsl Num) band 16#ffffffffffffffff) bor (A bsr (64 - Num)) end.

rotateRight(A) -> fun(Num) ->
  (A bsr Num) bor ((A bsl (64 - Num)) band 16#ffffffffffffffff) end.

subtract(A) -> fun(B) -> A - B end.

toBytes(A) -> fun(true) -> array:from_list (binary_to_list(<<A:64/little-integer>>));
                 (false) -> array:from_list (binary_to_list(<<A:64/big-integer>>))
              end.

toInt(A) -> A band 16#ffffffff.

toNumber(A) -> A * 1.0.

toSigned(A) -> A.

toString(A) -> fun(R) -> integer_to_binary (A, R) end.

toUnsigned(A) -> A.

xor(A) -> fun(B) -> A bxor B end.
